### PR TITLE
main/openrc: Allow http:// for ssh_key

### PR DESCRIPTION
**Pull request by mtarsel**
A slight revert of commit 4ef156921798938 which removed http://*
Since the kernel argument is downloading the public key, we
are not required to use an encrypted protocol. This will also
be easier for a local development environment.

### CODE REVIEW

**Comment by HRio** [comment]
its not for the encryption but for the authentication https is used... http will allow a man in the middle to inject other public keys (e.g. a few extra public keys).

**Comment by mtarsel** [comment]
Using http will not allow for a MITM attack since this patch is allowing us to send the _public key_ over the http connection. Looking at [RFC 4252 Section 7](https://www.ietf.org/rfc/rfc4252.txt), there is a signature based on the private key. Using public key authentication, the possession of a private key serves as authentication - not how the public key is transported.

**Comment by HRio** [comment]
we are retrieving a file, this file can contain more than one public key, i.e. a man in the middle attacker can modify this file and **add** public keys for private keys that the attacker is in possession off.

**Comment by tmh1999** [comment]
I support the point that https/ftps should be required, so that mitm could not inject his public key.
But I also support the point that for local-network installation, sometimes it is hard to setup https/ftps and http/ftp are just fine.

**Comment by HRio** [comment]
> easier for a local development environment

key url via https: https://github.com/mtarsel.keys (or any area on github where you can add stuff, but nobody else).

Or are we talking about an environment without Internet access or where github is not trusted?

**Comment by mtarsel** [comment]
Thanks, now I understand the attack method here. We should only send a file with 1 key because you're right, the attacker can add more keys, but the user shouldn't be sending a .pub with all their public keys in it anyways. Users would likely run `ssh-keygen` and use that single line .pub for their Alpine instance. Once an ssh session has been established, more keys could be added to grant more users access. Yes, as @tmh1999 states, allowing http would be much easier for local network install and so maybe adding a check to verify the ssh_key file is one line will suffice? Either a simple if statement in [init](https://github.com/alpinelinux/mkinitfs/blob/master/initramfs-init.in#L641) or add the 1 line check after the wget of the ssh_key file in firstboot.

**Comment by mtarsel** [comment]
> Or are we talking about an environment without Internet access or where github is not trusted?

I am talking about a local network install like PXE on a LAN which may not have external internet access. I don't think we should rely on a third party website to gain access after a local network install.

Another idea would be to check that the ssh_key file is only 1 line. Users would run `ssh-keygen` and use that single entry .pub for their Alpine instance. Once ssh session is established, more users could be granted ssh access. Either an If statement in [init](https://github.com/alpinelinux/mkinitfs/blob/master/initramfs-init.in#L641) or after the wget of the .pub to verify it is only 1 line.

**Comment by HRio** [comment]
limiting to one line will limit the use case, e.g. on sites with more than one admin (rather common) and where this feature is used in initial VM boot strapping.

**Comment by mtarsel** [comment]
A couple more ideas:

- Use a kernel command line argument like `http_enabled` or `sshkey_enable_http` but default to use https/ftps 
- Somehow include a checksum of the ssh_key file

**Comment by tmh1999** [comment]
I think ```sshkey_enable_http```, or ```ssh_key_http``` seems nice. @clandmeter

**Comment by tmh1999** [comment]
This should work : http://tpaste.us/7NJO
```
diff --git a/main/openrc/firstboot.initd b/main/openrc/firstboot.initd
index 4c26a815c8..34082e37f5 100644
--- a/main/openrc/firstboot.initd
+++ b/main/openrc/firstboot.initd
@@ -15,16 +15,21 @@ init_KOPT() {
 start() {
 	rm -f /etc/runlevels/*/$RC_SVCNAME
 	init_KOPT
-	local rc=0
+	local ssh_key="$KOPT_ssh_key" http="https" ftp="ftps" rc=0
 	ebegin "Starting ${RC_SVCNAME}"
-	if [ -n "$KOPT_ssh_key" ] && [ ! -f "/root/.ssh/authorized_keys" ]; then
+	if [ -n "$KOPT_ssh_key_plain" ]; then
+		ssh_key="$KOPT_ssh_key_plain"
+		http="http"
+		ftp="ftp"
+	fi
+	if [ -n "$ssh_key" ]  && [ ! -f "/root/.ssh/authorized_keys" ]; then
 		einfo "Fetching ssh keys"
 		mkdir -pm 700 /root/.ssh
-		case "$KOPT_ssh_key" in
-			https://*|ftps://*)
-				wget -q "$KOPT_ssh_key" -O /root/.ssh/authorized_keys
+		case "$ssh_key" in
+			"$http"://*|"$ftp"://*)
+				wget -q "$ssh_key" -O /root/.ssh/authorized_keys
 				rc=$?;;
-			*) echo "$KOPT_ssh_key" > /root/.ssh/authorized_keys;;
+			*) echo "$ssh_key" > /root/.ssh/authorized_keys;;
 		esac
 		chmod 600 /root/.ssh/authorized_keys
 	fi
diff --git a/initramfs-init.in b/initramfs-init.in
index d6f90a2..efd94e8 100755
--- a/initramfs-init.in
+++ b/initramfs-init.in
@@ -276,7 +276,7 @@ set -- $(cat /proc/cmdline)
 myopts="alpine_dev autodetect autoraid chart cryptroot cryptdm cryptheader cryptoffset
 	cryptdiscards cryptkey debug_init dma init_args keep_apk_new modules ovl_dev
 	pkgs quiet root_size root usbdelay ip alpine_repo apkovl alpine_start splash
-	blacklist overlaytmpfs rootfstype rootflags nbd resume s390x_net dasd ssh_key"
+	blacklist overlaytmpfs rootfstype rootflags nbd resume s390x_net dasd ssh_key ssh_key_plain"
 
 for opt; do
 	case "$opt" in
@@ -637,7 +637,7 @@ if [ "$KOPT_chart" = yes ]; then
 fi
 
 # add openssh
-if [ -n "$KOPT_ssh_key" ]; then
+if [ -n "$KOPT_ssh_key" ] || [ -n "$KOPT_ssh_key_plain" ]; then
 	pkgs="$pkgs openssh"
 	rc_add sshd default
 fi
```

**Comment by mtarsel** [comment]
thank you @tmh1999 I like the idea of adding `ssh_key_plain=http://` as a kernel command. LGTM

**Comment by mtarsel** [comment]
https://github.com/alpinelinux/mkinitfs/pull/29

**Comment by jpadams** [comment]
I'm fine with `ssh_key_plain` if that's the preferred way forward, but I find it much more elegant to just use `ssh_key` and allow the user to elect `http`, `https`, `ftps` depending on their own requirements. If you want to avoid MITM attacks, use `https`. If you want to be able to easily test things locally and it's safe, go ahead and use `http`.

**Comment by ncopa** [comment]
I applied this, but i should have fixed the commit message, which is wrong:
> Since the kernel argument is downloading the public key, we are not required to use an encrypted protocol.

As commented above, we still need https to prevent man-in-the-middle attacks.

The reason I committed it is that there are a few situations where plain text may be acceptable (LAN where you have full physical control of everything connected, eg virtual hosts, etc). There may also be situations where user understands and accept the risks involved with plain http and want allow that.

In other words, we will allow users to shoot themselves in the foot if they insist doing so.
